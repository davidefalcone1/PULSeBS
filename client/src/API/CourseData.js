//usata per la retrieve della lista dei corsi di un docente
class CourseData{
    constructor(courseId, courseName, teacherId, 
        normalBookingsAvgWeek, cancelledBookingsAvgWeek, waitingBookingsAvgWeek, 
        normalBookingsAvgMonth, cancelledBookingsAvgMonth, waitingBookingsAvgMonth,
        attendanceCountAvgMonth, attendanceCountAvgWeek ) {
        if (courseId) 
            this.courseId = courseId;
        this.courseName = courseName;
        this.teacherId = teacherId;
        this.normalBookingsAvgWeek = normalBookingsAvgWeek;
        this.cancelledBookingsAvgWeek = cancelledBookingsAvgWeek;
        this.waitingBookingsAvgWeek = waitingBookingsAvgWeek;
        this.normalBookingsAvgMonth = normalBookingsAvgMonth;
        this.cancelledBookingsAvgMonth = cancelledBookingsAvgMonth;
        this.waitingBookingsAvgMonth = waitingBookingsAvgMonth;
        this.attendanceCountAvgWeek = attendanceCountAvgWeek;
        this.attendanceCountAvgMonth = attendanceCountAvgMonth;
    }

    static fromJson(json){
        const temp =  Object.assign(new CourseData(), json);
        return temp;
    }
}

export default CourseData;