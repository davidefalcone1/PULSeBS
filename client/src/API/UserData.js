class UserData{
    constructor(id, personId, fullName, email) {
        if (id) 
            this.id = id;
        this.personId = personId;
        this.fullName = fullName;
        this.email = email;
    }

    static fromJson(json){
        const temp =  Object.assign(new UserData(), json);
        return temp;
    }
}

export default UserData;