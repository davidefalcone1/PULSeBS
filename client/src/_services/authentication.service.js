import { BehaviorSubject } from 'rxjs';


const currentUserSubject = new BehaviorSubject(JSON.parse(localStorage.getItem('user')));

export const authenticationService = {
    login,
    logout,
    user: currentUserSubject.asObservable(),
    get currentUserValue () { return currentUserSubject.value }
};

function handleResponse(response) {
    return response.text().then(text => {
        const data = text && JSON.parse(text);
        if (!response.ok) {
            if ([401, 403].indexOf(response.status) !== -1) {
                // auto logout if 401 Unauthorized or 403 Forbidden response returned from api
                authenticationService.logout();
            }

            const error = (data && data.message) || response.statusText;
            return Promise.reject(error);
        }

        return data;
    });
}

function login(username, password) {
    const requestOptions = {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ username, password })
    };
    return new Promise(async function (resolve, reject) {
        try {
            const response = await fetch('/users/authenticate', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({ username, password }),
            });
            const user = await response.json().then(handleResponse);
            if (response.ok) {
                localStorage.setItem('user', JSON.stringify(user));
                currentUserSubject.next(user);
                resolve(user);
            }
            else
                reject();
        }
        catch (e) {
            reject();
        }
    });

}
function logout() {

    const requestOptions = {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
    };
    localStorage.removeItem('user');
    currentUserSubject.next(null);
    return fetch('/logout',requestOptions);
}