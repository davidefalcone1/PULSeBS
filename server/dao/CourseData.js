//usata per la retrieve della lista dei corsi
class CourseData{
    constructor(courseId, courseName, teacherId) {
        if (courseId) 
            this.courseId = courseId;
        this.courseName = courseName;
        this.teacherId = teacherId;
    }
}

module.exports = CourseData;