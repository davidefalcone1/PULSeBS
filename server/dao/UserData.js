class UserData{
    constructor(id, personId, fullName, email) {
        if (id) 
            this.id = id;
        this.personId = personId;
        this.fullName = fullName;
        this.email = email;
    }

}

module.exports = UserData;