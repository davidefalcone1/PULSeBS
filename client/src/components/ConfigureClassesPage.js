import React from 'react';
import ConfigureClassItem from './ConfigureClassesItem';
import ListGroup from 'react-bootstrap/ListGroup';
import Button from 'react-bootstrap/esm/Button';
import Row from 'react-bootstrap/Row';
import Modal from 'react-bootstrap/Modal';
import Alert from 'react-bootstrap/Alert';
import Form from 'react-bootstrap/Form';
import { AuthContext } from '../_services/AuthContext';
import { Redirect } from 'react-router-dom';

const configureClassesPage = (props) => {
  return(
    <ConfigureClasses classesList = {props.classesList} createNewClass = {props.createNewClass}
      uploadFileClassrooms={props.uploadFileClassrooms}/>
  )
}

class ConfigureClasses extends React.Component {
  constructor(props) {
    super(props);
    this.props = props;
    this.state = {
      file: undefined, isUploading: false, errorFile: false,
      isCreating: false, classRoomName: '', maxSeats: 0,
      errorName: false, errorSeats: false
    }
  }

  activateModal = () => {
    this.setState({isCreating: true});
  }
  activateUploadFileModal = () => {
    this.setState({isUploading: true})
  }

  updateField = (name, value) => {
    this.setState({[name]: value}, () => {
      if(this.state.classRoomName !== ''){
        this.setState({errorName: false});
      }
      if(this.state.maxSeats > 0){
          this.setState({errorSeats: false});
      }
      if(this.state.file !== undefined && this.state.file !== ''){
        this.setState({errorFile: false});
      }
    });
  }

  handleSubmit = () => {
    if (!this.form.checkValidity()) {
        this.form.reportValidity();
    }
    else if(this.state.classRoomName === ''){
        this.setState({errorName: true});
    }
    else if(this.state.maxSeats <= 0){
        this.setState({errorSeats: true});
    }
    else {
        this.props.createNewClass(this.state.classRoomName, this.state.maxSeats)
    }
  }
  handleSubmitFile = () => {
    if (!this.formFile.checkValidity()) {
      this.formFile.reportValidity();
    }
    else if(this.state.file === undefined || this.state.file === ''){ 
      this.setState({errorFile: true});
    }
    else {
        this.props.uploadFileClassrooms(this.state.file)
        this.setState({isUploading: false})
    }
  }
  
  render(){
    return(
      <AuthContext.Consumer>
        {(context) => (
          <>        
            {context.user && this.props.classesList &&
              <>
                <br/>
                <Row className="justify-content-around">
                    <Button variant="primary" onClick={(event) => {
                          event.preventDefault();
                          this.activateModal();
                      }} id={"activateModalOfClasses"}>
                          Add New
                    </Button>
                    <Button variant="primary" onClick={(event) => {
                          event.preventDefault();
                          this.activateUploadFileModal();
                      }} id={"uploadFileOfClasses"}>
                          Upload File
                    </Button>
                </Row>
                <ListGroup as="ul" variant="flush">
                    <ListHeader />
                    {this.props.classesList.map((c) => 
                        <ConfigureClassItem key = {c.classId} class = {c}/>)}
                </ListGroup>

                <Modal show={this.state.isCreating} animation={false} scrollable={true}>
                  <Modal.Header>
                    <Modal.Title>Create new class</Modal.Title>
                  </Modal.Header>
                  <Modal.Body>
                    <Form method="POST" action="" id="newClassForm" onSubmit={(ev) => {
                      ev.preventDefault();
                      this.handleSubmit();
                    }} ref={(form) => this.form = form}>
                          
                      <Form.Group>
                        <Form.Label className="control-label">Classroom Name</Form.Label>
                        <Form.Control type="text" name="classRoomName" size = "lg"
                          value = {this.state.classRoomName} required autoFocus
                          onChange={(ev) => this.updateField(ev.target.name, ev.target.value)}/>
                      </Form.Group>
                      <Form.Group>
                        <Form.Label className="control-label">Max Available Seats</Form.Label>
                        <Form.Control type="number" name="maxSeats" size = "lg" required
                          value = {this.state.maxSeats} onChange={(ev) => 
                            this.updateField(ev.target.name, Number(ev.target.value))}/>
                      </Form.Group>
                      <Form.Group>
                        <div>
                          <button type="submit" className="btn btn-primary">Create</button>
                        </div>
                      </Form.Group>
                    </Form>
                  </Modal.Body>
                  <Modal.Footer>
                    <Button variant="secondary" type="button" 
                      onClick={(event) => {
                        event.preventDefault();
                        this.setState({isCreating: false});
                      }}>Close</Button>
                    {this.state.errorName &&
                      <>
                        <br/>
                        <Alert key="nameError" variant="danger">
                          Invalid name.
                        </Alert>
                      </>}
                    {this.state.errorSeats &&
                      <>
                        <br/>
                        <Alert key="seatsError" variant="danger">
                          Invalid seats number.
                        </Alert>
                      </>}
                  </Modal.Footer>
                </Modal>

                <Modal show={this.state.isUploading} animation={false} scrollable={true}>
                  <Modal.Header>
                    <Modal.Title>Upload file</Modal.Title>
                  </Modal.Header>
                  <Modal.Body>
                    <Form method="POST" action="" id="newClassFormFile" onSubmit={(ev) => {
                      ev.preventDefault();
                      this.handleSubmitFile();
                    }} ref={(form) => this.formFile = form}>                          
                      <Form.Group>
                        <Form.Label className="control-label">Insert file</Form.Label>
                        <Form.Control type="file" name="file" size = "lg"
                          required autoFocus accept=".csv"
                          onChange={(ev) => {                            
                            var f2 =function readFileContent(file) {
                              const reader = new FileReader()
                                return new Promise((resolve, reject) => {
                                  reader.onload=event=>resolve(event.target.result)
                                  reader.onerror = error => reject(error)
                                  reader.readAsText(file)
                                })
                            }
                            
                            f2(ev.target.files[0]).then(content => {
                              this.updateField("file", content)
                            }).catch(error => console.log(error))
                          }}/>
                      </Form.Group>
                      <Form.Group>
                        <div>
                          <button type="submit" className="btn btn-primary">
                            UPLOAD
                          </button>
                        </div>
                      </Form.Group>
                    </Form>
                  </Modal.Body>
                  <Modal.Footer>
                    <Button variant="secondary" type="button" 
                      onClick={(event) => {
                        event.preventDefault();
                        this.setState({isUploading: false});
                      }}>Close</Button>
                    {this.state.errorFile &&
                      <>
                        <br/>
                        <Alert key="fileError" variant="danger">
                          Invalid file.
                        </Alert>
                      </>}
                  </Modal.Footer>
                </Modal>
              </>
            }

            {context.user && !this.props.classesList &&
              <NoItemsImage/>
            }
            {!context.user && <Redirect to="/login"/>}
          </>
        )}
      </AuthContext.Consumer>
    );
  };  
}

function ListHeader() {
  return(
    <ListGroup.Item id = {"classesList-header"}>
        <div className="d-flex w-100 pt-3 justify-content-between no-gutters">
          <div className="col-sm-6">
              <h4>Class Name</h4>
          </div>
          <div className="col-sm-6">
              <h4>Max Available Seats</h4>
          </div>
        </div>
    </ListGroup.Item>
  );
}
function NoItemsImage(props){
  return(
      <div className="col-sm-12 pt-3">
          <img width="800" height="600" className="img-button" src='./images/no_data_image.png' alt=""/>
      </div>
  );
}

export default configureClassesPage;
