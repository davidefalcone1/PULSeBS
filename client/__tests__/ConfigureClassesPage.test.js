import React from 'react';
import renderer from 'react-test-renderer';
import { AuthContext } from '../src/_services/AuthContext';
import ConfigureClassesPage from '../src/components/ConfigureClassesPage';
import { BrowserRouter } from 'react-router-dom';
import CourseData from '../src/API/CourseData';
import UserData from '../src/API/UserData';

describe('ConfigureClassesPage', ()=>{
    test('Renders correctly', ()=>{
        const component = renderer.create(
            <BrowserRouter>
                <AuthContext.Provider value={{user: new UserData(1, '12345', 'Davide Falcone', 'davide.falcone@polito.it')}}>
                    <ConfigureClassesPage createNewEnrollment={()=>{}} uploadNewEnrollment={()=>{}} enrollmentInfos={[{courseId: 'XY123', studentId: '12345'}]} courses={[new CourseData('XY123', 'Software engineering 2', '654321', 2, 3, 5, 6, 7, 3, 5, 7)]} studentsInfos={[new UserData(1, '12345', 'Davide Falcone', 'davide.falcone@polito.it')]}/>
                </AuthContext.Provider>
            </BrowserRouter>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});