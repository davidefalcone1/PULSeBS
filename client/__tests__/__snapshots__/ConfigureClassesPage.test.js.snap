// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigureClassesPage Renders correctly 1`] = `
Array [
  <br />,
  <div
    className="justify-content-around row"
  >
    <button
      className="btn btn-primary"
      disabled={false}
      id="activateModalOfClasses"
      onClick={[Function]}
      type="button"
    >
      Add New
    </button>
    <button
      className="btn btn-primary"
      disabled={false}
      id="uploadFileOfClasses"
      onClick={[Function]}
      type="button"
    >
      Upload File
    </button>
  </div>,
  <ul
    className="list-group list-group-flush"
    onKeyDown={[Function]}
  >
    <div
      className="list-group-item"
      data-rb-event-key={null}
      disabled={false}
      id="classesList-header"
      onClick={[Function]}
    >
      <div
        className="d-flex w-100 pt-3 justify-content-between no-gutters"
      >
        <div
          className="col-sm-6"
        >
          <h4>
            Course
          </h4>
        </div>
        <div
          className="col-sm-6"
        >
          <h4>
            Student
          </h4>
        </div>
      </div>
    </div>
    <div
      className="list-group-item"
      data-rb-event-key={null}
      disabled={false}
      id="class-12345-XY123"
      onClick={[Function]}
    >
      <div
        className="d-flex w-100 pt-3 justify-content-between no-gutters"
      >
        <div
          className="col-sm-6"
        >
          <h6
            id="nameOfcourseForEnrollement_c-XY123-12345"
          >
            Software engineering 2
          </h6>
        </div>
        <div
          className="col-sm-6"
        >
          <h6
            id="nameOfstudentForEnrollement_s-XY123-12345"
          >
            Davide Falcone
          </h6>
        </div>
      </div>
    </div>
  </ul>,
]
`;
